File: vol1/ch7/LeetCode173-1.cpp
class BSTIterator {
private:
    int curp=-1;
    vector<int> nums;
    void inOrder(TreeNode* root) {
        if (root!=NULL) {
                inOrder(root->left);
                nums.push_back(root->val);
                inOrder(root->right);
        }
    }
public:
    BSTIterator(TreeNode* root) {
        inOrder(root);
    }

    int next() {
        return nums[++curp];
    }
    
    bool hasNext() {
        return curp!=nums.size()-1;
    }
};

File: vol1/ch7/LeetCode173-1.py
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.curp=-1
        self.nums=[]
        self.inorder(root)

    def inorder(self,root):
        if root!=None:
            self.inorder(root.left)
            self.nums.append(root.val)
            self.inorder(root.right)

    def next(self) -> int:
        self. curp +=1
        x=self.nums[self. curp]
        return x

    def hasNext(self) -> bool:
        return self. curp!=len(self.nums)-1
